# Đồ án KHDL cuối kỳ — Chuyên đề 3: Phân tích và gợi ý bất động sản TP.HCM

## Notebook 01: Mô tả bài toán và dữ liệu

- **Học viên**: [Nguyễn Minh Huy]
- **Môn**: Lập trình cho Khoa học Dữ liệu
- **Giảng viên hướng dẫn**: [PGS.TS Nguyễn Duy Hàm]
- **Phạm vi**: Căn hộ/Chung cư TP.HCM (6 quận nội/ngoại thành)

## 1. Bối cảnh và mục tiêu

Tin đăng bất động sản căn hộ/chung cư tại TP.HCM thường có:
- **Giá viết không thống nhất** ("3,19 tỷ", "2,1 tỷ", `price_text`).
- **Diện tích dạng số** (24-1323 m², có outlier rõ).
- **Hướng nhà và hướng ban công dạng mã số** (1..8 cho 8 hướng chính).
- **Tình trạng nội thất và pháp lý dạng mã** (furnishing_status 1..4, legal_status 1,2,4,5,6).
- **Nhiều tin thiếu thông tin** (hướng, nội thất, pháp lý) → cần imputation.

Mục tiêu đồ án:
1. Xây dựng **pipeline chuẩn hóa** district, direction code, furnishing, legal.
2. **Lọc outlier** và **recompute** `price_per_m2` để đảm bảo nhất quán.
3. **Dự đoán** `price_per_m2` từ các đặc trưng (diện tích, phòng, hướng, nội thất, pháp lý, quận, dự án).
4. **Phân khúc** bằng K-Means và **gợi ý top 5** theo hồ sơ nhu cầu.

## 2. Sáu câu hỏi nghiên cứu

1. Quận nào (trong 6 quận có dữ liệu) có giá trên mét vuông (`price_per_m2`) cao nhất?
2. Diện tích và số phòng ảnh hưởng thế nào đến tổng giá và giá/m²?
3. Hướng nhà, hướng ban công, tình trạng nội thất, pháp lý có liên hệ gì với giá không?
4. Tin nào có giá/diện tích bất thường hoặc có khả năng trùng?
5. Mô hình dự đoán giá/m² đạt sai số bao nhiêu (MAE / RMSE / R²)?
6. Top 5 tin nào phù hợp với từng hồ sơ nhu cầu (gia đình trẻ / nhà đầu tư / mua cao cấp)?

## 3. Hai nguồn dữ liệu

| Nguồn | File | Số dòng | Mô tả |
|---|---|---|---|
| 1 (chính) | `data/raw/real_estate_apartment.xlsx` | 1799 | Tin đăng BĐS căn hộ tại 6 quận TP.HCM, crawl từ chotot, schema 32 cột. |
| 2 (phụ) | `data/raw/neighborhood_amenities.csv` | ~89 | Thông tin tiện ích (trường học, bệnh viện, siêu thị, công viên, bus) theo (quận, phường) — tự tổng hợp deterministic. |

## 4. Từ điển dữ liệu (rút gọn)

| Cột | Kiểu | Mô tả |
|---|---|---|
| `listing_id` | int | Mã tin đăng (unique). |
| `title`, `description` | str | Tiêu đề + mô tả chi tiết (nhiều emoji, có thông tin giá trị). |
| `property_type` | str | Loại hình (toàn bộ là "Căn hộ/Chung cư"). |
| `province` | str | Tỉnh/TP (toàn bộ là "Tp Hồ Chí Minh"). |
| `district` | str | Quận (6 giá trị — đã chuẩn: Quận Bình Thạnh, Quận Gò Vấp, Quận 12, Thành phố Thủ Đức, Quận 7, Quận Bình Tân). |
| `ward`, `street` | str | Phường/xã (83 ward) và tên đường. |
| `latitude`, `longitude` | float | Toạ độ thực (17 missing). |
| `area_m2` | float | Diện tích (m²), có 1 outlier 1323 m². |
| `bedrooms`, `bathrooms` | float | Số phòng. |
| `direction`, `balcony_direction` | float | Mã hướng 1..8 (�ông/Tây/Nam/Bắc/Đông Nam/Tây Nam/Đông Bắc/Tây Bắc). |
| `furnishing_status` | float | Mã nội thất 1..4 (không/cơ bản/đầy đủ/cao cấp). |
| `legal_status` | float | Mã pháp lý 1,2,4,5,6 (đang cập nhật/sổ hồng lâu dài/H�MB/sổ hồng chung/sổ hồng riêng). |
| `apartment_type` | int | Mã loại căn hộ (1..6). |
| `total_price` | int | Tổng giá (VND). |
| `price_text` | str | Giá dạng text ("3,19 tỷ"). |
| `price_per_m2` | float | Giá/m² tính sẵn (sẽ tính lại). |
| `image_count` | int | Số ảnh đăng kèm. |
| `posted_at`, `updated_at`, `crawled_at` | datetime | Ngày đăng/cập nhật/crawl. |

Chi tiết từng cột xem `reports/data_dictionary.md`.

## 5. Dữ liệu mẫu — 5 dòng đầu

In [1]:
import pandas as pd

raw = pd.read_excel('data/raw/real_estate_apartment.xlsx')
print('Shape:', raw.shape)
raw.head()

Shape: (1799, 32)


,id,listing_id,title,description,property_type,apartment_type,project_name,province,district,ward,...,price_per_m2,posted_at,updated_at,image_count,status,source,url,crawled_at,created_at,db_updated_at
0,2925,177735723,👉 Chung Cư Bình Thạnh - 56m² - Thang Máy -Nội ...,"✅ Chung cư 234 Phan Văn Trị, hẻm 330 Phan Văn ...",Căn hộ/Chung cư,1,NaN,Tp Hồ Chí Minh,Quận Bình Thạnh,Phường 12,...,57.0,2026-08-04 02:03:40.236,2026-07-28 05:16:52,8,active,chotot,https://www.chotot.com/13000/1010/177735723.htm,2026-08-08 08:27:11.843,2026-08-08 08:27:11.843,2026-08-08 08:44:55.888
1,2926,177944633,"Chợ Cây Quéo giáp Gò Vấp ,CH 51m2 sổ sẵn ,vay ...",Dự án: \nThông tin chi tiết: CH cổ điển Nhật B...,Căn hộ/Chung cư,1,NaN,Tp Hồ Chí Minh,Quận Bình Thạnh,Phường 11,...,41.0,2026-08-08 06:29:04.900,2026-08-06 09:12:02,8,active,chotot,https://www.chotot.com/13000/1010/177944633.htm,2026-08-08 08:27:11.957,2026-08-08 08:27:11.957,2026-08-08 08:50:59.990
2,2927,177687427,"CH Bình Thạnh 2,1 tỷ /2PN /thuê 13tr/th,sổ sẵn...","- DT: 52m2 –2PN , ngân hàng hỗ trợ vay 25 năm\...",Căn hộ/Chung cư,1,NaN,Tp Hồ Chí Minh,Quận Bình Thạnh,Phường 12,...,40.0,2026-08-08 01:50:07.250,2026-07-26 02:15:30,5,active,chotot,https://www.chotot.com/13000/1010/177687427.htm,2026-08-08 08:27:12.068,2026-08-08 08:27:12.068,2026-08-08 08:50:00.189
3,2928,177431899,"Căn 2PN Lê Quang Định ở liền nhà y hình, sổ sẵ...",- Cần tiền bán gấp căn hộ có sẵn nội thất vào ...,Căn hộ/Chung cư,1,NaN,Tp Hồ Chí Minh,Quận Bình Thạnh,Phường 5,...,41.0,2026-08-07 01:22:54.974,2026-07-14 07:17:55,7,active,chotot,https://www.chotot.com/13000/1010/177431899.htm,2026-08-08 08:27:12.129,2026-08-08 08:27:12.129,2026-08-08 08:27:12.129
4,2929,176746606,"Căn góc 2,05 tỷ/ 2PN (sổ sẵn) vay 80% _ gần C...","Vị trí: Ngay trục Phạm Văn Đồng, sát cầu Bình ...",Căn hộ/Chung cư,1,NaN,Tp Hồ Chí Minh,Quận Bình Thạnh,Phường 5,...,41.0,2026-08-06 22:34:53.578,2026-07-14 07:08:49,6,active,chotot,https://www.chotot.com/13000/1010/176746606.htm,2026-08-08 08:27:12.189,2026-08-08 08:27:12.189,2026-08-08 08:27:12.189


## 6. Dữ liệu mẫu — 5 dòng cuối

In [2]:
raw.tail()

,id,listing_id,title,description,property_type,apartment_type,project_name,province,district,ward,...,price_per_m2,posted_at,updated_at,image_count,status,source,url,crawled_at,created_at,db_updated_at
1794,6273,174331297,tecco 2 phòng ngủ cần bán dt58m2 gía chỉ từ 2....,Cần bán căn hộ tecco\n2 phòng ngủ 2 vệ sinh\nd...,Căn hộ/Chung cư,1,Tecco Green Nest,Tp Hồ Chí Minh,Quận 12,Phường Tân Thới Nhất,...,38.0,2026-07-29 09:53:47.000,NaT,7,active,chotot,https://www.chotot.com/13000/1010/174331297.htm,2026-08-08 09:17:13.966,2026-08-08 09:17:13.966,2026-08-08 09:17:13.966
1795,6276,177467828,"GIÁ RẤT TỐT 1PN VIEW MÁT RƯỢT, K ỒN Q7 RIVERSI...","✔ Diện tích 53m², thiết kế 1PN rộng rãi, tối ư...",Căn hộ/Chung cư,1,Saigon Riverside Complex Q7,Tp Hồ Chí Minh,Quận 7,Phường Phú Thuận,...,58.0,2026-07-16 01:24:20.000,NaT,9,active,chotot,https://www.chotot.com/13000/1010/177467828.htm,2026-08-08 09:17:14.137,2026-08-08 09:17:14.137,2026-08-08 09:17:14.137
1796,6296,177456742,Bán căn hộ lô góc 2PN+1 65.7m2 giá 3.76 tỷ,❗️DEAL SIÊU HỜI❗️CĂN HỘ LÔ GÓC 2PN+1 ĐA NĂNG N...,Căn hộ/Chung cư,1,NaN,Tp Hồ Chí Minh,Quận 7,Phường Phú Mỹ,...,57.0,2026-07-15 08:13:16.000,NaT,3,active,chotot,https://www.chotot.com/13000/1010/177456742.htm,2026-08-08 09:17:15.895,2026-08-08 09:17:15.895,2026-08-08 09:17:15.895
1797,6312,177475251,Bán căn hộ 1pn Ascentia sẵn hợp đồng thuê 아센티아...,"Bán gấp 1pn Ascentia\nDiện tích 49m2, 1PN 1WC...",Căn hộ/Chung cư,1,The Ascentia Phú Mỹ Hưng,Tp Hồ Chí Minh,Quận 7,Phường Tân Phú,...,120.0,2026-07-16 06:07:14.000,NaT,3,active,chotot,https://www.chotot.com/13000/1010/177475251.htm,2026-08-08 09:17:17.485,2026-08-08 09:17:17.485,2026-08-08 09:17:17.485
1798,6315,177449230,"chính chủ cần bán cc Phú Mỹ 93m2,Quận 7 kế bên...","Chào bán căn cc Phú Mỹ, kế bên Phú Mỹ Hưng và ...",Căn hộ/Chung cư,1,Chung cư Phú Mỹ,Tp Hồ Chí Minh,Quận 7,Phường Phú Mỹ,...,57.0,2026-07-17 05:53:04.137,2026-07-15 04:28:59,10,active,chotot,https://www.chotot.com/13000/1010/177449230.htm,2026-08-08 09:17:17.665,2026-08-08 09:17:17.665,2026-08-08 09:17:17.665


## 7. Schema và dtype

In [3]:
raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 1799 entries, 0 to 1798
Data columns (total 32 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   id                 1799 non-null   int64         
 1   listing_id         1799 non-null   int64         
 2   title              1799 non-null   str           
 3   description        1799 non-null   str           
 4   property_type      1799 non-null   str           
 5   apartment_type     1799 non-null   int64         
 6   project_name       1423 non-null   str           
 7   province           1799 non-null   str           
 8   district           1799 non-null   str           
 9   ward               1783 non-null   str           
 10  street             1789 non-null   str           
 11  latitude           1782 non-null   float64       
 12  longitude          1782 non-null   float64       
 13  area_m2            1793 non-null   float64       
 14  bedrooms           

## 8. Tỉ lệ missing theo cột

In [4]:
(raw.isna().mean() * 100).round(2).sort_values(ascending=False)

direction            78.93
balcony_direction    73.15
updated_at           48.80
furnishing_status    34.35
project_name         20.90
legal_status         20.01
bathrooms            15.56
latitude              0.94
longitude             0.94
ward                  0.89
street                0.56
price_per_m2          0.33
area_m2               0.33
bedrooms              0.17
image_count           0.00
status                0.00
source                0.00
url                   0.00
crawled_at            0.00
created_at            0.00
posted_at             0.00
id                    0.00
price_text            0.00
total_price           0.00
listing_id            0.00
district              0.00
province              0.00
apartment_type        0.00
property_type         0.00
description           0.00
title                 0.00
db_updated_at         0.00
dtype: float64

## 9. Kết luận

Dữ liệu 1799 dòng, 32 cột, phạm vi 6 quận TP.HCM. Các cột có missing cao (`direction` 79%, `balcony_direction` 73%, `furnishing_status` 34%, `legal_status` 20%) sẽ được Pipeline `SimpleImputer` xử lý trong notebook 04. Cột có dtype `float` cho các categorical (direction, furnishing, legal) → cần decode sang text cho EDA nhưng giữ mã số cho ML.

Tiếp theo: mở `02_collection_and_cleaning.ipynb` để xem pipeline làm sạch chi tiết.